wrappers/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,568 @@
+		pkcs11.NewAttribute(pkcs11.CKA_ID, keyIdBytes),
+		pkcs11.NewAttribute(pkcs11.CKA_DECRYPT, true),
+	}
+	if kms.label != "" {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_LABEL, []byte(kms.label)))
+	}
+	if kms.mechanism != 0 {
+		keyTypeString, err := GetKeyTypeFromMech(kms.mechanism)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get key type from mechanism: %s", err)
+		}
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, keyTypeString))
+	}
+	if err := p.FindObjectsInit(session, template); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+
+	obj, _, err := p.FindObjects(session, 2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjects: %s", err)
+	}
+	if err := p.FindObjectsFinal(session); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 FindObjectsFinal: %s", err)
+	}
+
+	if len(obj) != 1 {
+		return nil, fmt.Errorf("expected 1 object, got %d", len(obj))
+	}
+	key := obj[0]
+
+	template = []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil),
+	}
+	attr, err := p.GetAttributeValue(session, pkcs11.ObjectHandle(key), template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 GetAttributeValue: %s", err)
+	}
+
+	attrMap := GetAttributesMap(attr)
+	keyType := GetValueAsInt(attrMap[pkcs11.CKA_KEY_TYPE])
+
+	mechanism := uint(0)
+	switch keyType {
+	case pkcs11.CKK_AES:
+		if kms.mechanism != 0 {
+			mechanism = kms.mechanism
+		} else {
+			mechanism = pkcs11.CKM_AES_CBC_PAD
+		}
+	case pkcs11.CKK_RSA:
+		if kms.mechanism != 0 {
+			mechanism = kms.mechanism
+		} else {
+			mechanism = pkcs11.CKM_RSA_PKCS
+		}
+	default:
+		return nil, fmt.Errorf("unsupported key type: %d", keyType)
+	}
+
+	var iv []byte
+	if IsIvNeeded(mechanism) {
+		template = []*pkcs11.Attribute{
+			pkcs11.NewAttribute(pkcs11.CKA_VALUE_LEN, nil),
+		}
+		attr, err := p.GetAttributeValue(session, pkcs11.ObjectHandle(key), template)
+		if err != nil {
+			return nil, fmt.Errorf("failed to pkcs11 GetAttributeValue: %s", err)
+		}
+		attrMap := GetAttributesMap(attr)
+
+		ivLength := 0
+		ivLength = int(GetValueAsInt(attrMap[pkcs11.CKA_VALUE_LEN]))
+
+		if len(encryptedDEK) < ivLength {
+			return nil, fmt.Errorf("encrypted DEK is too short")
+		}
+
+		iv = encryptedDEK[:ivLength]
+		encryptedDEK = encryptedDEK[ivLength:]
+	}
+
+	if err = p.DecryptInit(session, []*pkcs11.Mechanism{pkcs11.NewMechanism(mechanism, iv)}, key); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 DecryptInit: %s", err)
+	}
+
+	var decrypted []byte
+	if decrypted, err = p.Decrypt(session, encryptedDEK); err != nil {
+		return nil, fmt.Errorf("failed to pkcs11 Decrypt: %s", err)
+	}
+	return decrypted, nil
+}
+
+func GetAttributesMap(attrs []*pkcs11.Attribute) map[uint][]byte {
+	m := make(map[uint][]byte, len(attrs))
+	for _, a := range attrs {
+		m[a.Type] = a.Value
+	}
+	return m
+}
+
+func GetValueAsInt(value []byte) int64 {
+	if value == nil {
```

Comment:
This shouldn't need to happen as `len(nil) == 0` and we return the same value rather than erring.

Reply:
I removed the unnecessary conditional statements.